### PR TITLE
upgrade syntax to Python 3.7

### PR DESCRIPTION
**Pull request by basnijholt**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] All tests passing
- [ ] I have added tests to cover my changes
- [ ] I have updated documentation (if necessary)
- [ ] My code follows the code style of this project
(we are using `black`: you can `pip install pre-commit`,
run `pre-commit install` in the `pydra` directory
and `black` will be run automatically with each commit)

## Acknowledgment
- [x] I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

File: pydra/engine/helpers_state.py [hunk 1]
```diff
@@ -17,7 +17,7 @@ def __init__(self, value):
         self.value = value
 
     def __str__(self):
-        return "{}".format(self.value)
+        return f"{self.value}"
```
Comment (effigies) [inline pydra/engine/helpers_state.py:20]: This seems like a weird one to use an f-string for.

```suggestion
        return str(self.value)
```

File: pydra/engine/helpers_state.py [hunk 2]
```diff
@@ -425,8 +425,7 @@ def splits(splitter_rpn, inputs, inner_inputs=None, cont_dim=None):
     # analysing states from connected tasks if inner_inputs
     if inner_inputs:
         previous_states_ind = {
-            "_{}".format(v.name): (v.ind_l_final, v.keys_final)
-            for _, v in inner_inputs.items()
+            f"_{v.name}": (v.ind_l_final, v.keys_final) for _, v in inner_inputs.items()
```
Comment (effigies) [inline pydra/engine/helpers_state.py:428]:
```suggestion
            f"_{v.name}": (v.ind_l_final, v.keys_final) for v in inner_inputs.values()
```

File: pydra/engine/helpers_state.py [hunk 3]
```diff
@@ -576,7 +575,7 @@ def splits_groups(splitter_rpn, combiner=None, inner_inputs=None):
         combiner = []
     if inner_inputs:
         previous_states_ind = {
-            "_{}".format(v.name): v.keys_final for _, v in inner_inputs.items()
+            f"_{v.name}": v.keys_final for _, v in inner_inputs.items()
```
Comment (effigies) [inline pydra/engine/helpers_state.py:578]:
```suggestion
            f"_{v.name}": v.keys_final for v in inner_inputs.values()
```

File: pydra/engine/task.py [hunk 4]
```diff
@@ -625,7 +625,7 @@ def binds(self, opt, ind=None):
         """
         bargs = []
         for (key, val) in self.bind_paths(ind).items():
-            bargs.extend([opt, "{0}:{1}:{2}".format(key, val[0], val[1])])
+            bargs.extend([opt, "{}:{}:{}".format(key, val[0], val[1])])
```
Comment (effigies) [inline pydra/engine/task.py:628]:
```suggestion
            bargs.extend([opt, f"{key}:{val[0]}:{val[1]}"])
```

File: pydra/engine/tests/data_tests/loading.py [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 
 def loading(filename):
-    with open(filename, "r") as f:
+    with open(filename) as f:
```
Comment (effigies) [inline pydra/engine/tests/data_tests/loading.py:5]: Explicit is generally better than implicit. I would revert this one:

```suggestion
    with open(filename, "r") as f:
```

File: pydra/engine/tests/test_dockertask.py [hunk 6]
```diff
@@ -1059,10 +1057,10 @@ def test_docker_cmd_inputspec_copyfile_1(plugin, tmpdir):
     assert res.output.out_file.exists()
     # the file is  copied, and than it is changed in place
     assert res.output.out_file.parent == docky.output_dir
-    with open(res.output.out_file, "r") as f:
+    with open(res.output.out_file) as f:
```
Comment (effigies) [inline pydra/engine/tests/test_dockertask.py:1060]: Looks like there are a lot of these...

File: pydra/utils/profiler.py [hunk 7]
```diff
@@ -86,7 +86,10 @@ def _sample(self, cpu_interval=None):
             except psutil.NoSuchProcess:
                 pass
 
-        print("%f,%f,%f,%f" % (time(), cpu, rss / _MB, vms / _MB), file=self._logfile)
+        print(
+            "{:f},{:f},{:f},{:f}".format(time(), cpu, rss / _MB, vms / _MB),
```
Comment (effigies) [inline pydra/utils/profiler.py:90]:
```suggestion
            f"{time():f},{cpu:f},{rss / _MB:f},{vms / _MB:f}",
```

File: pydra/engine/tests/data_tests/loading.py (hunk 5 above)
Comment (basnijholt) [reply pydra/engine/tests/data_tests/loading.py:5]: I feel this one is pretty obvious.

I also don't see `d.get(key, None)` everywhere for example 😄 :trollface:

Comment (djarecka) [reply pydra/engine/tests/data_tests/loading.py:5]: I use often `with open(filename, "r") as f`, but I agree that ` with open(filename) as f` is pretty standard and looks good to me

**Comment by djarecka** [comment]
@basnijholt - thank you for the PR! The only tests that are failing are the ones that checks the style. This is because we are using `pre-commit` to run `black` with each commit

**Comment by basnijholt** [comment]
I am just leaving for two weeks of vacation. So unfortunately I cannot run black now. I did give you guys push permission on this branch, if you want to fix it before merging. Otherwise you could just push one commit up on top of master after the merge.

**Comment by djarecka** [comment]
If you want ton be included in the zenodo file, please edit the [file](https://github.com/nipype/pydra/blob/master/.zenodo.json) or provide the info, so I can add you

**Pull request merged**